### PR TITLE
Added Counting Sort [Java]

### LINKED ISSUE #101: Counting Sort [Java]
A popular algorithm used for sorting under specific constraints but if constraints were followed, It is the one  of the fastest sort

www.geeksforgeeks.org/counting-sort/

### PR DESCRIPTION
**Pull request by Hsankesara**
Fixed #101 

#### By submitting this pull request I confirm I've read and complied with the below requirements.

- [x] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [x] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [x] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.

### CODE REVIEW

File: counting_sort/CountingSort.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import java.util.Random;
+ class CountingSort {
+
```
Comment (Monal5031) [inline counting_sort/CountingSort.java]: Please remove unnecessary blank spaces.

File: counting_sort/CountingSort.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+		int A[]=new int[10000];
+        int k=0;
+        Random rand=new Random();
+        for(int i=0;i<A.length;i++) {
+            A[i]=rand.nextInt(100);
+            if(k<A[i])
+            	k=A[i];
+            }
+        
+        A=cSort(A,k);
```
Comment (Monal5031) [inline counting_sort/CountingSort.java]: space after comma `cSort(A, k);`

File: counting_sort/CountingSort.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import java.util.Random;
+ class CountingSort {
+	
+	public static void main(String[] args) {
+		
+		int A[]=new int[10000];
```
Comment (Monal5031) [inline counting_sort/CountingSort.java]: Attach brackets with type specifier `int[] A` not `int A[]`

File: counting_sort/CountingSort.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        int k=0;
+        Random rand=new Random();
+        for(int i=0;i<A.length;i++) {
+            A[i]=rand.nextInt(100);
+            if(k<A[i])
+            	k=A[i];
+            }
+        
+        A=cSort(A,k);
+        for(int i=0;i<A.length;i++)
```
Comment (Monal5031) [inline counting_sort/CountingSort.java]: space between for and ( as `for (`

File: counting_sort/CountingSort.java (hunk 3 above)
Comment (Monal5031) [inline counting_sort/CountingSort.java]: Please Correct indentations.

File: counting_sort/CountingSort.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        	System.out.print(A[i] + " ");
+	}
+
+	public static int[] cSort(int[] a,int k) {
```
Comment (Monal5031) [inline counting_sort/CountingSort.java]: Rename method to filename to best extent possible as mentioned in contributing.md
You can use `countingSort()`

File: counting_sort/CountingSort.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+		int[] b=new int[a.length];
+		int[] c= new int[k+1];
+		 getC(a,c);
+        for(int i=a.length-1;i>=0;i--){
```
Comment (Monal5031) [inline counting_sort/CountingSort.java]: Space before and after binary operators. `i >= 0`

**Comment by mohitkyadav** [comment]
@Hsankesara You should use a bottom-up approach.

File: counting_sort/CountingSort.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
+import java.util.Random;
+ class CountingSort {
+	public static void main(String[] args) {
+		int[] A=new int[10000];
+        int k=0;
+        Random rand=new Random();
+        for(int i=0;i<A.length;i++) {
```
Comment (Monal5031) [inline counting_sort/CountingSort.java]: Space before and after binary operators.
Make this change for whole code.

File: counting_sort/CountingSort.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
+import java.util.Random;
+ class CountingSort {
+	public static void main(String[] args) {
+		int[] A=new int[10000];
+        int k=0;
```
Comment (Monal5031) [inline counting_sort/CountingSort.java]: Please correct the indentations.

**Comment by Monal5031** [comment]
@Hsankesara I agree with @mohitkyadav Bottom-up approach would be nice.

File: counting_sort/CountingSort.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+	public static void main(String[] args) {
+		int[] A = new int[10000];
+        	int k = 0;
+        	Random rand=new Random();
+        	for (int i = 0;i < A.length;i++) {
+            		A[i] = rand.nextInt(100);
+            		if(k < A[i])
+            			k = A[i];
+            	}
+       		 A = countingSort(A, k);
```
Comment (Monal5031) [inline counting_sort/CountingSort.java]: Intended wrong, please remove 1 tab space

**Comment by Hsankesara** [comment]
@Monal5031  Intend is fine , I checked it and it is 1 tabspace , I don't know where did you get that misconception???

**Comment by Monal5031** [comment]
@Hsankesara It is clearly visible, please take a look in files changed section you will get it what I'm trying to say. Just see and compare the indentations of `countOccurences()` and rest of the methods.

File: counting_sort/CountingSort.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+	 public static void main(String[] args) {
+		int[] A = new int[10000];
+        	int k = 0;
+        	Random rand=new Random();
+        	for (int i = 0;i < A.length;i++) {
+            		A[i] = rand.nextInt(100);
+            		if(k < A[i])
+            			k = A[i];
+          	}
+       		 A = countingSort(A, k);
```
Comment (Monal5031) [inline counting_sort/CountingSort.java]: This line went off the desired indentation rest of the code looks good to me.

**Comment by Hsankesara** [comment]
Now it is correct @Monal5031 , isn't it?

**Comment by Hsankesara** [comment]
@aviaryan  @singhpratyush @yashLadha please review the PR so I can work on the next one...

File: counting_sort/CountingSort.java (hunk 10 above)
Comment (Monal5031) [inline counting_sort/CountingSort.java]: @Hsankesara have a look over this.

Comment (Hsankesara) [inline counting_sort/CountingSort.java]: It is closed bracket of  for loop and seems fine in place, what'd you expect to change @Monal5031 ?

Comment (aviaryan) [inline counting_sort/CountingSort.java]: `A = countingSort(A, k);`

Comment (aashutoshrathi) [inline counting_sort/CountingSort.java]: @Hsankesara there is a extra space before ` A = countingSort(A, k);` make it `A = countingSort(A, k);`

Comment (Monal5031) [inline counting_sort/CountingSort.java]: @Hsankesara I am talking about `A = countingSort(A, k);` not the closing bracket of for loop.

Comment (Hsankesara) [inline counting_sort/CountingSort.java]: It's Done @Monal5031 , and sorry for misinterpretation

Comment (Monal5031) [inline counting_sort/CountingSort.java]: @Hsankesara Its okay, happens many times with me too.

File: counting_sort/CountingSort.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
+import java.util.Random;
+ class CountingSort {
```
Comment (singhpratyush) [inline counting_sort/CountingSort.java]: Correct alignment issue and add a line break before.

File: counting_sort/CountingSort.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+	private static void countOccurences(int[] a, int[] c) {
+		for(int i = 0;i < a.length;i++)
+           		c[a[i]]++;
+		for(int i = 1;i < c.length;i++)
+           		c[i] = c[i-1] + c[i];
+    	}
+	public static int[] countingSort(int[] a, int k) {
+		int[] b = new int[a.length];
+		int[] c = new int[k+1];
+		 countOccurences(a, c);
```
Comment (singhpratyush) [inline counting_sort/CountingSort.java]: Correct alignment issue.

File: counting_sort/CountingSort.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
+import java.util.Random;
+ class CountingSort {
+	private static void countOccurences(int[] a, int[] c) {
+		for(int i = 0;i < a.length;i++)
```
Comment (singhpratyush) [inline counting_sort/CountingSort.java]: Add space after `for` and semicolon. Correct similar instances.

File: counting_sort/CountingSort.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+		int[] b = new int[a.length];
+		int[] c = new int[k+1];
+		 countOccurences(a, c);
+        	for (int i = a.length-1;i >= 0;i--) {
+            		b[c[a[i]]-1] = a[i];
+            		c[a[i]]--;
+        	}       
+        	return b;	
+	}
+	 public static void main(String[] args) {
```
Comment (singhpratyush) [inline counting_sort/CountingSort.java]: Correct alignment issue.

File: counting_sort/CountingSort.java [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+		int[] A = new int[10000];
+        	int k = 0;
+        	Random rand=new Random();
```
Comment (singhpratyush) [inline counting_sort/CountingSort.java]: Add space.

**Comment by Hsankesara** [comment]
Changes done @singhpratyush , waiting for your response...

File: counting_sort/CountingSort.java (hunk 11 above)
Comment (singhpratyush) [inline counting_sort/CountingSort.java]: > add a line break before.

File: counting_sort/CountingSort.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
+import java.util.Random;
+class CountingSort {	//Time complexity = O(n) , but every number must be between 0 to k
+	private static void countOccurences(int[] a, int[] c) {	//Basically counting occurences of particular  number and store in its index
+		for (int i = 0; i < a.length; i++)	//Counting occurence of each number
+           		c[a[i]]++;
+		for(int i = 1;i < c.length;i++)		//counting total number of numbers less then or equal to that number
```
Comment (singhpratyush) [inline counting_sort/CountingSort.java]: Space after semicolon. Correct all occurances.

File: counting_sort/CountingSort.java [hunk 17]
```diff
@@ -0,0 +1,32 @@
+import java.util.Random;
+class CountingSort {	//Time complexity = O(n) , but every number must be between 0 to k
```
Comment (singhpratyush) [inline counting_sort/CountingSort.java]: Add line break before.

Comment (singhpratyush) [inline counting_sort/CountingSort.java]: Waiting for this to be fixed.

File: counting_sort/CountingSort.java [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+	 
+	public static int[] countingSort(int[] a, int k) {
+		int[] b = new int[a.length];
+		int[] c = new int[k+1];
+		 countOccurences(a, c);
+        	for (int i = a.length-1;i >= 0;i--) {
+            		b[c[a[i]]-1] = a[i];
+            		c[a[i]]--;
+        	}       
+        return b;
```
Comment (PiyushPawar17) [inline counting_sort/CountingSort.java]: Add a tab

**Comment by singhpratyush** [comment]
@Monal5031: Have you verified the functioning of this code?

**Comment by Monal5031** [comment]
@singhpratyush Yes sir I had verified it when he uploaded the code.

**Comment by singhpratyush** [comment]
Thank you @Hsankesara for your contribution. We look for more from you.

**Pull request merged**